### PR TITLE
Add gguf support for StableLM

### PR DESCRIPTION
**Pull request by VladOS95-cyber**
# What does this PR do?
Add GGUF support for StableLM

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Link: https://github.com/huggingface/transformers/issues/33260
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
Regarding the task @SunMarc @LysandreJik @ArthurZucker .

### CODE REVIEW

**Comment by VladOS95-cyber** [comment]
Hi @SunMarc! This PR is ready for review, please, take a look.

File: src/transformers/models/gpt_neox/tokenization_gpt_neox_fast.py [hunk 1]
```diff
@@ -105,8 +105,8 @@ def __init__(
         **kwargs,
     ):
         super().__init__(
-            vocab_file,
-            merges_file,
+            vocab_file=vocab_file,
+            merges_file=merges_file,
```
Comment (SunMarc) [inline src/transformers/models/gpt_neox/tokenization_gpt_neox_fast.py:109]: why did you add that ?

File: tests/quantization/ggml/test_ggml.py [hunk 2]
```diff
@@ -44,6 +44,7 @@ class GgufIntegrationTests(unittest.TestCase):
     phi3_model_id = "microsoft/Phi-3-mini-4k-instruct-gguf"
     bloom_model_id = "afrideva/bloom-560m-GGUF"
     original_bloom_model_id = "bigscience/bloom-560m"
+    stablelm_model_id = "afrideva/stablelm-3b-4e1t-GGUF"
```
Comment (SunMarc) [inline tests/quantization/ggml/test_ggml.py:47]: just like bloom, can you also add a fp16 test to see if we get the same result with the transformers model and the converted gguf model ?

**Review by SunMarc — COMMENTED** [review]
Thanks for adding @VladOS95-cyber as always ! Left a few comments

File: src/transformers/models/gpt_neox/tokenization_gpt_neox_fast.py (hunk 1 above)
Comment (VladOS95-cyber) [reply src/transformers/models/gpt_neox/tokenization_gpt_neox_fast.py:109]: Because of tokenization_utils_fast.py (L120-123). If we do not explicitly assign vocab_file and merges_file params just like that, `kwargs.get("vocab_file")` returns None and `gguf_param = load_gguf_checkpoint(kwargs.get("vocab_file"))` will not work. The same things we did for Bloom and Qwen

File: tests/quantization/ggml/test_ggml.py (hunk 2 above)
Comment (VladOS95-cyber) [reply tests/quantization/ggml/test_ggml.py:47]: Yes, agree, I just added

**Comment by VladOS95-cyber** [comment]
Hey @SunMarc! I resolved all comments, you could take a look again

**Review by SunMarc — COMMENTED** [review]
Thanks for iterating ! Could you also add a test to check the weights of each layer for the fp16 model, to see if we get the same model ? We recently merged the gguf falcon model and he added a nice test for that. Also there are a few merge conflits, thanks !

File: tests/quantization/ggml/test_ggml.py [hunk 3]
```diff
@@ -503,6 +508,75 @@ def test_falcon7b_weights_conversion_fp16(self):
...
+            torch_dtype=torch.float16,
+        )
+
+        converted_model = AutoModelForCausalLM.from_pretrained(
+            self.stablelm2_model_id,
+            gguf_file=self.fp16_stablelm2_model_id,
+            torch_dtype=torch.float16,
+            # for precise comparison it is required to use the original model config
+            # as quantized one is different in parameters: use_parallel_residual and use_qkv_bias
+            # and it highly influences on the output results
```
Comment (SunMarc) [inline tests/quantization/ggml/test_ggml.py:538]: I'm not use to understand why we need to use the original model config ? Since we are using the fp16 model, it shouldn't be quantized.

**Review by SunMarc — APPROVED** [review]
LGTM ! Left a question to better understand

File: tests/quantization/ggml/test_ggml.py (hunk 3 above)
Comment (VladOS95-cyber) [reply tests/quantization/ggml/test_ggml.py:538]: Because even in fp16 model we use config from gguf file and it is quite different compared with original model config. For example, `use_parallel_residual = True` in quantazied one, while in original model it is false. If it is true, the models skips creating post_attention_layernorm. And `use_qkv_bias = False` in quantized one, and True in original model. If it is false, it skips using q, k, v biases. That's why we explicitly use original model config in order to have all weights and biases correctly created for fair comparison.

Comment (SunMarc) [reply tests/quantization/ggml/test_ggml.py:538]: Thanks for the explanation ! This makes sense to me. The issue is on the conversion script from hf to gguf then ? If one specify fp16, it shoudn't modify `use_parallel_residual` and `use_qkv_bias`.

Comment (VladOS95-cyber) [reply tests/quantization/ggml/test_ggml.py:538]: So, I suppose yes and no, use_parallel_residual = True is specified in conversion script by default `self.gguf_writer.add_parallel_residual(hparams["use_parallel_residual"] if "use_parallel_residual" in hparams else True)`, `use_qkv_bias` is not specified at all, so by default it is False according to config implementation.

Comment (SunMarc) [reply tests/quantization/ggml/test_ggml.py:538]: Oh thx ! That will be very helpful for the next models we add !

**Comment by VladOS95-cyber** [comment]
HI @LysandreJik! Just a kind reminder to take a look on this PR!

**Review by LysandreJik — APPROVED** [review]
Thanks @VladOS95-cyber, this seems to have passed through the net. Looks great to me!
